wallet-new/src/Cardano/Wallet/Kernel/Wallets.hs

```diff
@@ -142,3 +184,81 @@ createWalletHdRnd pw hasSpendingPassword name assuranceLevel esk utxo = do
             if hasSpendingPassword then HD.HasSpendingPassword created
                                    else HD.NoSpendingPassword
 
+deleteHdWallet :: PassiveWallet
+               -> HD.HdRootId
+               -> IO (Either HD.UnknownHdRoot ())
+deleteHdWallet wallet rootId = do
+    -- STEP 1: Remove the HdRoot via an acid-state transaction which will
+    --         also delete any associated accounts and addresses.
+    res <- update' (wallet ^. wallets) $ DeleteHdRoot rootId
+    case res of
+        Left err -> return (Left err)
+        Right () -> do
+            -- STEP 2: Purge the key from the keystore.
+            Keystore.delete (WalletIdHdRnd rootId) (wallet ^. walletKeystore)
+            return $ Right ()
+
+{-------------------------------------------------------------------------------
+  Wallet update
+-------------------------------------------------------------------------------}
+
+updateHdWallet :: PassiveWallet
+               -> HD.HdRootId
+               -> HD.AssuranceLevel
+               -> HD.WalletName
+               -> IO (Either HD.UnknownHdRoot (Kernel.DB, HdRoot))
+updateHdWallet pw hdRootId assuranceLevel walletName = do
+    res <- update' (pw ^. wallets) (UpdateHdWallet hdRootId assuranceLevel walletName)
+    case res of
+         Left e              -> return (Left e)
+         Right (db, newRoot) -> return (Right (db, newRoot))
+
+updatePassword :: PassiveWallet
+               -> HD.HdRootId
+               -> PassPhrase
+               -- ^ The old 'PassPhrase' for this Wallet.
+               -> PassPhrase
+               -- ^ The new 'PassPhrase' for this Wallet.
+               -> IO (Either UpdateWalletPasswordError (Kernel.DB, HdRoot))
+updatePassword pw hdRootId oldPassword newPassword = do
+    let keystore = pw ^. walletKeystore
+        wId = WalletIdHdRnd hdRootId
+    -- STEP 1: Lookup the key from the keystore
+    mbKey <- Keystore.lookup wId keystore
+    case mbKey of
+         Nothing -> return $ Left $ UpdateWalletPasswordKeyNotFound hdRootId
+         Just esk -> do
+             -- STEP 2: Check that the 2 password matches. While this in
+             --         principle could be checked on the wallet layer side,
+             --         it's preferrable to do everything in the kernel to make
+             --         this operation really atomic.
```

Comment:
I don't understand this comment. What is atomic about this? You do the check and then you do the replacement. Nothing stopping something else from interfering in the middle right?

Reply:
Oh, not atomic is the concurrency sense, bad wording, sorry. I wanted to convey that I wanted to have the kernel function do everything related to this functionality rather than having the piece of logic scattered between the layer and the kernel. Will fix the wording.

Reply:
Nonetheless we _may_ want some atomicity here. What if the password gets changed someplace else after you check it?